Add id2label support, fix CUDA bug in return_probabilities

takamai06 opened the pull request:
### What is the issue?

This issue proposes three enhancements and fixes for the KServe project:

---

~### 1. **Expose Model Pull Resources in Helm Chart**~

~The current `kserve-resources` Helm chart does not allow customization of the resource requirements used for pulling models. In environments with large models or tight node constraints, this limitation can cause pod scheduling issues or failures.~

**~Proposed Change~**: 
~Add configurable values in the Helm chart to control CPU and memory requests/limits used during model pulling.~

---

### 2. **Add `id2label` Support in `huggingfaceserver`**

Currently, the HuggingFace server does not support returning human-readable labels mapped from IDs, even if the `id2label` mapping exists in the model config.

**Proposed Change**:  
Add support for an optional `id2label` argument:
- If enabled, it will return the mapped label for each prediction.
- When combined with `return_probabilities`, it will return both the label and the score.

---

### 3. **Fix `return_probabilities` Crash on CUDA**

There is a bug in the HuggingFace server when running with CUDA:
- The postprocess method tries to convert a tensor to NumPy before moving it to CPU.
- This results in a `500 Internal Server Error`.

**Proposed Fix**:  
Ensure `.cpu()` is called before converting tensors to NumPy in the postprocessing logic.

---

### Why is this needed?

These changes:
- Improve configurability in deployment environments.
- Enhance model output interpretability for classification tasks.
- Fix a critical bug preventing GPU inference from working with probability outputs.

HotsauceLee committed:
Fix: Isvc matched with wrong ModelCache (#4398)

Signed-off-by: Gavin Li <yli1109@bloomberg.net>
Co-authored-by: Dan Sun <dsun20@bloomberg.net>
Signed-off-by: tohana <tohana@akamai.com>

huazq committed:
Fix: remove duplicated OpenAIGenerativeModel in __init__ (#4399)

Signed-off-by: cilivel.hua <cilivel.hua@gmail.com>
Co-authored-by: Dan Sun <dsun20@bloomberg.net>
Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
feat: refactor storage initializer resources configuration

Updated the storage initializer in the Helm chart to use dynamic resource configuration from values.yaml, allowing for more flexible resource management.

Signed-off-by: tohana <tohana@akamai.com>

tylerhyang committed:
Feat: Allow inference service metadata injection at agent sidecar level for payload logging (#4325)

Signed-off-by: Jooho Lee <jlee@redhat.com>
Signed-off-by: Tyler Yang <tyler.yang@cloudera.com>
Signed-off-by: Tyler Yang <69994946+tylerhyang@users.noreply.github.com>
Co-authored-by: Jooho Lee <jlee@redhat.com>
Co-authored-by: Tyler Yang <tyler.yang@cloudera.com>
Co-authored-by: Filippe Spolti <filippespolti@gmail.com>
Signed-off-by: tohana <tohana@akamai.com>

andyi2it committed:
Fix for KEDA scaledobject target value is set to pointer instead of specified value (#4373)

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>
Co-authored-by: Dan Sun <dsun20@bloomberg.net>
Signed-off-by: tohana <tohana@akamai.com>

Jooho committed:
MultiNode change a logic to calculate ray node count and gpu count (#4356)

Signed-off-by: tohana <tohana@akamai.com>

greenmoon55 committed:
Remove internal annotations when no cache resource is matched (#4412)

Signed-off-by: Jin Dong <jdong183@bloomberg.net>
Signed-off-by: tohana <tohana@akamai.com>

israel-hdez committed:
Add DeploymentMode to InferenceService and InferenceGraph status and prevent deploymentMode change (#4423)

Signed-off-by: Vedant Mahabaleshwarkar <vmahabal@redhat.com>
Signed-off-by: Dan Sun <dsun20@bloomberg.net>
Signed-off-by: Edgar Hernández <23639005+israel-hdez@users.noreply.github.com>
Co-authored-by: Vedant Mahabaleshwarkar <vmahabal@redhat.com>
Co-authored-by: Dan Sun <dsun20@bloomberg.net>
Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
feat: add LabelTransformer class for label mapping and processing

- Introduced a new LabelTransformer class that loads label mappings from a config file and processes inference responses to convert predictions into human-readable labels.
- Updated the main execution to instantiate LabelTransformer instead of ImageTransformer.

fix: handle tensor device compatibility in HuggingfaceEncoderModel

- Modified the handling of probabilities and outputs in the HuggingfaceEncoderModel to ensure compatibility with both CPU and GPU tensors by moving them to CPU before conversion to numpy.

Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
revert: did some commit that should not be done

Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
feat: add --use_id2label argument for classification outputs

This commit introduces a new command-line argument `--use_id2label` to enable the use of the model's id2label mapping for classification outputs. The argument is added to the argument parser and is integrated into the model loading and inference processes to enhance output readability.

Signed-off-by: tohana <tohana@akamai.com>

johnzheng1975 committed:
Feat should not NewHPARecroncile for external hpa. (#4363)

Signed-off-by: John Zheng <john.zheng@hp.com>
Signed-off-by: tohana <tohana@akamai.com>

sivanantha321 committed:
chore: Upgrade prow-github-actions to version 2 in workflow files (#4417)

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>
Signed-off-by: tohana <tohana@akamai.com>

Jooho committed:
Fix: update external autoscaler tests missed from PR 4363 (#4438)

Signed-off-by: Jooho Lee <jlee@redhat.com>
Signed-off-by: tohana <tohana@akamai.com>

terrytangyuan committed:
Update OWNERS (#4442)

Signed-off-by: Yuan Tang <terrytangyuan@gmail.com>
Signed-off-by: tohana <tohana@akamai.com>

sivanantha321 commented on charts/kserve-resources/templates/clusterstoragecontainer.yaml:
```diff
@@ -7,12 +7,9 @@ spec:
-      limits:
-        memory: 1Gi
-        cpu: "1"
+      {{- with .Values.kserve.storage.resources }}
```
Can we leave this change as there is separate PR for this ?

takamai06 replied:
yes, i will do that

sivanantha321 commented on python/huggingfaceserver/huggingfaceserver/encoder_model.py:
```diff
@@ -302,24 +305,37 @@ def postprocess(
+                        probs = probs.cpu()  # Move to CPU for dict conversion which only works on CPU
+                    if self.use_id2label and hasattr(self.model_config, 'id2label'):
+                        inferences.append({self.model_config.id2label[i]: float(prob) for i, prob in enumerate(probs[0])})
+                    else:
```
Probably add a warning message that model config does not have id2label field.

sivanantha321 commented:
@takamai06 Please add a test case for this.

takamai06 committed:
Revert "feat: refactor storage initializer resources configuration"

This reverts commit 9dfe0154d3c998f73e8936c9946cbe6c59a77a0c.

Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
refactor: simplify inference label handling

- Introduced a helper method `_get_label_or_index` to streamline the process of retrieving labels from the `id2label` mapping or returning the index when necessary. This reduces code duplication and improves readability in the inference logic.

Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
feat(tests): add unit tests for encoder label mapping functionality

This commit introduces a new test file that contains unit tests for the `_get_label_or_index` helper method in the `HuggingfaceEncoderModel`. The tests cover various scenarios, including cases with and without label mappings, handling of probabilities, and edge cases for label mapping.

Signed-off-by: tohana <tohana@akamai.com>

takamai06 committed:
docs: update README to include usage of --use_id2label flag for classification tasks

Added instructions for using human-readable labels instead of indices in classification tasks by adding the `--use_id2label` flag. Included examples for both classification outputs with and without this flag.

Signed-off-by: tohana <tohana@akamai.com>

takamai06 commented:
@sivanantha321 i added all the things you asked :)

gavrissh commented:
@takamai06 You would need to sign off the PR to fix the DCO build

takamai06 committed:
Merge remote-tracking branch 'origin/master' into id2label

takamai06 committed:
Merge branch 'master' into id2label

takamai06 commented:
> @takamai06 You would need to sign off the PR to fix the DCO build

already done that :) 
@gavrissh can you approve?

takamai06 committed:
Merge branch 'master' into id2label

gavrissh reviewed (approved):
/lgtm

Thanks for the contribution

takamai06 committed:
Merge branch 'master' into id2label

sivanantha321 committed:
Merge branch 'master' into id2label

sivanantha321 commented:
Can you help resolve the formatting issue by running `make precommit`

It would be nice if you can help update the website docs as well ! https://github.com/kserve/website/blob/main/docs/modelserving/v1beta1/huggingface/text_classification/README.md

takamai06 committed:
style: format code for better readability and consistency

Refactored code to improve readability by adjusting spacing and line breaks, ensuring consistent formatting throughout the `encoder_model.py` and `test_encoder_label_mapping.py` files.

Signed-off-by: tohana <tohana@akamai.com>

takamai06 commented:
@sivanantha321  done. regarding the website docs i will do it later :) 

> Can you help resolve the formatting issue by running `make precommit`
> 
> It would be nice if you can help update the website docs as well ! https://github.com/kserve/website/blob/main/docs/modelserving/v1beta1/huggingface/text_classification/README.md

gavrissh commented:
Should we merge this one?

sivanantha321 committed:
Merge branch 'master' into id2label

sivanantha321 commented:
>   Full diff:
  - [{0: -2.152204 ± 9.0e-06, 1: 2.5094059 ± 9.0e-06}]
  + [{0: 0.009362765587866306, 1: 0.9906372427940369}]

@takamai06 The e2e test is failing due to output mismatch

oleksii-donets commented:
@takamai06 @sivanantha321 @gavrissh
Are there any plans to finalize this PR?
My team has run into a limitation where id2label is not supported, so having it available out of the box would be very beneficial and would help us avoid implementing a custom runtime solution.
